src/jit/importer.cpp

```diff
@@ -18269,4 +18276,286 @@ bool Compiler::IsMathIntrinsic(GenTreePtr tree)
+    // used to infer struct-ness.
+    //
+    // A rough accounting in corelib shows this breakdown for node types
+    // seen as 'this' obj in vtable calls is as follows:
+    //
+    // Node Type  Count  HasType
+    // lclVar     4256      3825
+    // field      1098      1098
+    // retExp      864
+    // intrin      318
+    // helper      294
+    // callv-vtbl  224
+    // []          123
+    // call         61
+    // box          60
+    // call-nchk    27
+    // callv-stub    5
+    //
+    // Note some lclVars are missing class handles; presumably these
+    // are locals introduced by the compiler. We might be able to add
+    // in type info for them. The table above gives a rough priority
+    // order for adding type support.
+    //
+    // Even when we can deduce the type, we may not be able to
+    // devirtualize, but if we can't deduce the type, we can't do
+    // anything.
+    //
+    // In corelib, devirtualization happens roughly 5% of the time,
+    // that is, out of 8014 virtual call sites, we devirtualize at
+    // 453. Assuming we could fill in missing types, devirtualization
+    // might ultimately kick in for around ~10% of virtual call sites.
+    //
+    // To get beyond that we'd need to do something speculative,
+    // either by adding runtime type tests into the jitted code, or by
+    // adding the ability to pitch and replace code to safely undo
+    // assumptions made about class hierarchy.
+    CORINFO_CLASS_HANDLE objClass = nullptr;
+    const genTreeOps     objOp    = obj->OperGet();
+
+    switch (objOp)
+    {
+        case GT_LCL_VAR:
+        {
+            const unsigned objLcl = obj->AsLclVar()->GetLclNum();
+
+            objClass = lvaTable[objLcl].lvClassHnd;
+            break;
+        }
+
+        case GT_FIELD:
+        {
+            CORINFO_FIELD_HANDLE fieldHnd = obj->gtField.gtFldHnd;
+
+            if (fieldHnd != nullptr)
+            {
+                CORINFO_CLASS_HANDLE fieldClass   = nullptr;
+                CorInfoType          fieldCorType = info.compCompHnd->getFieldType(fieldHnd, &fieldClass);
+                if (fieldCorType == CORINFO_TYPE_CLASS)
+                {
+                    objClass = fieldClass;
+                }
+            }
+
+            break;
+        }
+
+        default:
+        {
+            break;
+        }
+    }
+
+    // Bail if we can't figure out the type of 'this'.
+    if (objClass == nullptr)
+    {
+        return;
+    }
+
+    // Fetch VM handles for base method and base class.
+    CORINFO_METHOD_HANDLE baseMethod = callInfo->hMethod;
+    CORINFO_CLASS_HANDLE  baseClass  = info.compCompHnd->getMethodClass(baseMethod);
+
+    // Bail out if there's missing class info.
+    if (baseClass == nullptr)
+    {
+        return;
+    }
+
+    // If the objClass is sealed (final), then we may be able to devirtualize.
+    const DWORD objClassAttribs = info.compCompHnd->getClassAttribs(objClass);
```

Comment:
Thinking of leaving things as they are now -- a few checks (eg the one for __Canon) were moved to the VM side, but a number remain on the jit side -- notably the attribute checks for final/sealed methods and classes.

Seemed logical to me to keep the policy decisions on the jit side and have the VM deal with correctness/servicing, which more or less dictates the current setup. Otherwise the jit has to pass in a bit more state (for instance, is the type known exactly) and get back more state (whether the target is known exactly).

Some of the jit-side checks will be removed in future work, eg interface call devirtualization.